### DISCUSSION TITLE
Automatically remove code owners as reviewers

### DISCUSSION THREAD

**Discussion by bjacobs09** [community/community#66936] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

We are assigning code owners. The PR then requires a review from that group.
But if the file that triggered it is removed from the PR, the required reviewer is not removed.
The original reviewer who is no longer needed should be automatically removed.

**Comment by Dipanshu089** [comment]
There are a few ways to work around this issue:

You can manually remove the required reviewer from the pull request.
You can use a third-party tool that integrates with GitHub, such as CodeTriage, to automatically remove required reviewers when files are removed from pull requests.
You can change the GitHub settings for your repository so that code owner reviews are not required for pull requests that remove files.
Here are the steps on how to manually remove a required reviewer from a pull request on GitHub:

Go to the pull request page.
Click on the "Reviewers" button.
In the list of reviewers, find the reviewer you want to remove and uncheck the box next to their name.
Click on the "Save" button.

**Reply by martin-honek-tg** [reply]
Sure, it is possible to remove reviewers manually - just as it is possible to add them manually. The point of assigning code owners is to automate such manual actions. While removing reviewers is much rarer than adding reviewers, it would be nice to have it implemented in the future.

**Reply by torressam333** [reply]
Is there a way to disable the option to require review from code owners? I still want 1 codeowners pr approval to be required but not auto email blast the codeowners for EVERY pr. How can I do this?

**Reply by dipakparmar** [reply]
I came here for the same reason. It's noisy; when the code owner is a team, it results in email spam. There should be a better way to handle this.

**Comment by olekxandr** [comment]
It seems like you are dealing with a common issue for code review with GitHub or similar version control systems. When a file is removed from a PR (Pull Request), the code owner assigned to the file should automatically be removed from the required reviewers list. However, this isn't always the case.

Let's distinguish between the two settings you are dealing with:

Code ownership: A feature where specific files or portions of a codebase can be assigned to a specific team or individual, such that when changes are made to those sections, the assigned "owner" will be notified and asked to review the corresponding PR.

Required reviewers: Often set up as part of branch protection rules in order to require approval from certain individuals or teams before changes can be merged.

It's important to remember that these mechanisms don't directly communicate. Herein lies your problem: even if a file is removed from a PR, a once-required review isn't necessarily removed.

To automatically remove a no-longer-needed reviewer, you must have a system in place that actively checks for changes and updates the required reviewers accordingly.

Unfortunately, there's no out-of-the-box functionality in GitHub or similar platforms that would do this for you. However, you could potentially implement such a system through a GitHub Actions workflow or another CI/CD tool.

You would need to set up a system that listens for events like 'synchronized' (which signals that new commits were pushed to the PR), and then checks if the changes still touch files owned by the currently assigned reviewers. If not, it uses the GitHub API to remove the reviewers.

Here's a rough sketch of how such a system could be implemented:

A webhook set up to listen for 'pull_request' events is triggered when changes are pushed to a PR.
The system fetches the PR details, including the list of files changed and the list of assigned reviewers.
The system checks if any of the files owned by the current reviewers are still among the changed files.
If not, it uses the GitHub API DELETE /repos/{owner}/{repo}/pulls/{pull_number}/requested_reviewers method to remove the reviewer.
Remember that you'll also need to handle API rate limits, errors, and edge cases to make this solution robust.

This would require creating a custom application or bot - it's not a trivial task but it is achievable with a fair understanding of programming concepts. I would recommend using a relevant language such as JavaScript (with Octokit library for the GitHub API), Python (with PyGithub library), or any other language you prefer. You'll want to familiarize yourself with GitHub webhooks and the GitHub API if you haven't worked with them before.
